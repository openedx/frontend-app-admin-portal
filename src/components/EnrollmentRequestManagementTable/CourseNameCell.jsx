import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Skeleton from 'react-loading-skeleton';
import {
  OverlayTrigger,
  Popover,
  Button,
  Hyperlink,
} from '@edx/paragon';
import { getConfig } from '@edx/frontend-platform/config';

import { useCourseDetails } from './data/hooks';

const CourseDetailsPopoverContentBase = ({ enterpriseSlug, courseKey }) => {
  const { ENTERPRISE_LEARNER_PORTAL_URL } = getConfig();
  const [courseDetails, isCourseDetailsLoading] = useCourseDetails(courseKey);

  if (isCourseDetailsLoading) {
    return (
      <>
        <Skeleton count={2} />
        <span className="sr-only">Loading course details...</span>
      </>
    );
  }

  return (
    <div>
      {(courseDetails?.shortDescription) && (
        <>
          <p>{courseDetails.shortDescription}</p>
          <hr />
        </>
      )}
      <p>
        <Hyperlink
          target="_blank"
          destination={`${ENTERPRISE_LEARNER_PORTAL_URL}/${enterpriseSlug}/course/${courseKey}`}
        >
          Learn more about this course
        </Hyperlink>
      </p>
    </div>
  );
};

CourseDetailsPopoverContentBase.propTypes = {
  enterpriseSlug: PropTypes.string.isRequired,
  courseKey: PropTypes.string.isRequired,
};

const mapStateToProps = (state) => ({
  enterpriseSlug: state.portalConfiguration.enterpriseSlug,
});

const CourseDetailsPopoverContent = connect(mapStateToProps)(CourseDetailsPopoverContentBase);

const CourseNameCell = ({ row }) => (
  <OverlayTrigger
    trigger="click"
    placement="top"
    rootClose
    overlay={(
      <Popover id="popover-requests-table-course-details">
        <Popover.Title as="h5">{row.original.courseName}</Popover.Title>
        <Popover.Content>
          <CourseDetailsPopoverContent courseKey={row.original.courseKey} />
        </Popover.Content>
      </Popover>
    )}
  >
    <Button
      variant="link"
      className="text-left px-0"
      size="sm"
    >
      {row.original.courseName}
    </Button>
  </OverlayTrigger>
);

CourseNameCell.propTypes = {
  row: PropTypes.shape({
    original: PropTypes.shape({
      courseName: PropTypes.string,
      courseKey: PropTypes.string,
    }).isRequired,
  }).isRequired,
};

export default connect(mapStateToProps)(CourseNameCell);
