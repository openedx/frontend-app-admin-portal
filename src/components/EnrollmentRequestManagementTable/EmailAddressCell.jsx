import React from 'react';
import PropTypes from 'prop-types';

const EmailAddressCell = ({ row }) => (
  <span data-hj-suppress>{row.original.emailAddress}</span>
);

EmailAddressCell.propTypes = {
  row: PropTypes.shape({
    original: PropTypes.shape({
      emailAddress: PropTypes.string,
    }),
  }).isRequired,
};

export default EmailAddressCell;
