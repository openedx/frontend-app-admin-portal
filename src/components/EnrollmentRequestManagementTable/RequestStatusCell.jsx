import PropTypes from 'prop-types';
import { Badge } from '@edx/paragon';

const capitalizeFirstLetter = string => string.charAt(0).toUpperCase() + string.slice(1);

const RequestStatusCell = ({ row }) => {
  const variant = row.original.requestStatus === 'requested' ? 'primary' : 'light';
  return (
    <Badge variant={variant}>
      {capitalizeFirstLetter(row.original.requestStatus)}
    </Badge>
  );
};

RequestStatusCell.propTypes = {
  row: PropTypes.shape({
    original: PropTypes.shape({
      requestStatus: PropTypes.string,
    }),
  }).isRequired,
};

export default RequestStatusCell;
