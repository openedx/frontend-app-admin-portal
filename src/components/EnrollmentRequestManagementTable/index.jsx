import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import {
  DataTable,
  TextFilter,
  CheckboxFilter,
} from '@edx/paragon';

import EmailAddressCell from './EmailAddressCell';
import RequestDateCell from './RequestDateCell';
import RequestStatusCell from './RequestStatusCell';
import CourseNameCell from './CourseNameCell';
import ActionCell from './ActionCell';

const EnrollmentRequestManagementTable = ({
  onApprove,
  onDecline,
  data,
  fetchData,
  requestStatusFilterChoices,
}) => {
  const columns = useMemo(
    () => ([
      {
        Header: 'Email address',
        accessor: 'emailAddress',
        Cell: EmailAddressCell,
      },
      {
        Header: 'Course name',
        accessor: 'courseName',
        Cell: CourseNameCell,
        disableFilters: true,
      },
      {
        Header: 'Request date',
        accessor: 'requestDate',
        Cell: RequestDateCell,
        disableFilters: true,
      },
      {
        Header: 'Request status',
        accessor: 'requestStatus',
        Cell: RequestStatusCell,
        Filter: CheckboxFilter,
        filter: 'includesValue',
        filterChoices: requestStatusFilterChoices,
      },
      {
        Header: '',
        accessor: 'action',
        disableFilters: true,
        Cell: (props) => (
          <ActionCell
            {...props}
            onApprove={onApprove}
            onDecline={onDecline}
          />
        ),
      },
    ]),
    [onApprove, onDecline, requestStatusFilterChoices],
  );

  return (
    <DataTable
      isFilterable
      manualFilters
      isPaginated
      manualPagination
      defaultColumnValues={{ Filter: TextFilter }}
      itemCount={3}
      fetchData={fetchData}
      data={data}
      pageCount={1}
      columns={columns}
    >
      <DataTable.TableControlBar />
      <DataTable.Table />
      <DataTable.EmptyTable content="No results found" />
      <DataTable.TableFooter />
    </DataTable>
  );
};

EnrollmentRequestManagementTable.propTypes = {
  fetchData: PropTypes.func.isRequired,
  data: PropTypes.arrayOf(PropTypes.shape({
    emailAddress: PropTypes.string,
    courseName: PropTypes.string,
    courseKey: PropTypes.string,
    requestDate: PropTypes.string,
    requestStatus: PropTypes.oneOf(['requested', 'declined']),
  })).isRequired,
  requestStatusFilterChoices: PropTypes.arrayOf(PropTypes.shape({
    number: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    name: PropTypes.string,
    value: PropTypes.string,
  })).isRequired,
  onApprove: PropTypes.func.isRequired,
  onDecline: PropTypes.func.isRequired,
};

export default EnrollmentRequestManagementTable;
