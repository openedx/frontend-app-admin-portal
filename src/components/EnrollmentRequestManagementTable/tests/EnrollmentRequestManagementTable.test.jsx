import React from 'react';
import renderer from 'react-test-renderer';

import EnrollmentRequestManagementTable from '..';

const defaultProps = {
  fetchData: jest.fn(),
  data: [{
    emailAddress: 'test@example.com',
    courseName: 'edX Demonstration Course',
    courseKey: 'edX+DemoX',
    requestDate: '2019-12-03T21:39:24.395101Z',
    requestStatus: 'requested',
  }, {
    emailAddress: 'test@example.com',
    courseName: 'edX Demonstration Course',
    courseKey: 'edX+DemoX',
    requestDate: '2019-12-03T21:39:24.395101Z',
    requestStatus: 'requested',
  }, {
    emailAddress: 'test@example.com',
    courseName: 'edX Demonstration Course',
    courseKey: 'edX+DemoX',
    requestDate: '2019-12-03T21:39:24.395101Z',
    requestStatus: 'declined',
  }],
  requestStatusFilterChoices: [{
    name: 'requested',
    number: 2,
    value: 'requested',
  }, {
    name: 'declined',
    number: 1,
    value: 'declined',
  }],
  onApprove: jest.fn(),
  onDecline: jest.fn(),
};

describe('EnrollmentRequestManagementTable', () => {
  test('renders data in a table as expected', () => {
    const tree = renderer
      .create(<EnrollmentRequestManagementTable {...defaultProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
