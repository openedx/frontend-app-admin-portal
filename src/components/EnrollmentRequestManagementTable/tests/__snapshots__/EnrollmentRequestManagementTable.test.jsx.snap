// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnrollmentRequestManagementTable renders data in a table as expected 1`] = `
<div
  className="pgn__data-table-layout-wrapper"
>
  <div
    className="pgn__data-table-layout-main"
  >
    <div
      className="pgn__data-table-wrapper"
    >
      <div
        className="pgn__data-table-status-bar"
      >
        <div
          className="pgn__data-table-actions"
        >
          <div
            className="pgn__data-table-actions-left"
          >
            <div
              className="pgn__data-table-filters"
            >
              <div
                className="mr-2"
              >
                <div
                  className="pgn__form-group"
                >
                  <label
                    className="pgn__form-label sr-only"
                    htmlFor="form-field2"
                    id="text-filter-label-header_emailAddress-1"
                  >
                    Search Email address
                  </label>
                  <input
                    aria-labelledby="text-filter-label-header_emailAddress-1"
                    className="form-control"
                    onChange={[Function]}
                    placeholder="Search Email address"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                className="dropdown"
              >
                <button
                  aria-expanded={false}
                  aria-haspopup={true}
                  className="dropdown-toggle btn btn-outline-primary"
                  disabled={false}
                  id="table-filters-dropdown"
                  onClick={[Function]}
                  type="button"
                >
                  Filters
                </button>
              </div>
            </div>
          </div>
          <div
            className="pgn__data-table-actions-right"
          >
            <div
              className="pgn__data-table-toggle"
            />
            <div />
          </div>
        </div>
        <div
          className="pgn__data-table-status"
        >
          <div
            className="pgn__data-table-status-left"
          >
            <div
              className="pgn__smart-status"
            >
              Showing 
              3
               of 
              3
              .
            </div>
          </div>
        </div>
      </div>
      <div
        className="pgn__data-table-container"
      >
        <table
          className="pgn__data-table is-striped"
          role="table"
        >
          <thead>
            <tr
              role="row"
            >
              <th
                colSpan={1}
                role="columnheader"
              >
                <span
                  className="d-flex align-items-center"
                >
                  <span>
                    Email address
                  </span>
                </span>
              </th>
              <th
                colSpan={1}
                role="columnheader"
              >
                <span
                  className="d-flex align-items-center"
                >
                  <span>
                    Course name
                  </span>
                </span>
              </th>
              <th
                colSpan={1}
                role="columnheader"
              >
                <span
                  className="d-flex align-items-center"
                >
                  <span>
                    Request date
                  </span>
                </span>
              </th>
              <th
                colSpan={1}
                role="columnheader"
              >
                <span
                  className="d-flex align-items-center"
                >
                  <span>
                    Request status
                  </span>
                </span>
              </th>
              <th
                colSpan={1}
                role="columnheader"
              >
                <span
                  className="d-flex align-items-center"
                >
                  <span>
                    
                  </span>
                </span>
              </th>
            </tr>
          </thead>
          <tbody
            role="rowgroup"
          >
            <tr
              className="pgn__data-table-row"
              role="row"
            >
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <span
                  data-hj-suppress={true}
                >
                  test@example.com
                </span>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <button
                  className="text-left px-0 btn btn-link btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  edX Demonstration Course
                </button>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                December 3, 2019
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <span
                  className="badge badge-primary"
                >
                  Requested
                </span>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <div
                  className="pgn__action-row"
                >
                  <button
                    className="btn btn-tertiary btn-inline"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    Decline
                  </button>
                  <button
                    className="btn btn-link btn-inline"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    Approve
                  </button>
                </div>
              </td>
            </tr>
            <tr
              className="pgn__data-table-row"
              role="row"
            >
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <span
                  data-hj-suppress={true}
                >
                  test@example.com
                </span>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <button
                  className="text-left px-0 btn btn-link btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  edX Demonstration Course
                </button>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                December 3, 2019
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <span
                  className="badge badge-primary"
                >
                  Requested
                </span>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <div
                  className="pgn__action-row"
                >
                  <button
                    className="btn btn-tertiary btn-inline"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    Decline
                  </button>
                  <button
                    className="btn btn-link btn-inline"
                    disabled={false}
                    onClick={[Function]}
                    type="button"
                  >
                    Approve
                  </button>
                </div>
              </td>
            </tr>
            <tr
              className="pgn__data-table-row"
              role="row"
            >
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <span
                  data-hj-suppress={true}
                >
                  test@example.com
                </span>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <button
                  className="text-left px-0 btn btn-link btn-sm"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  edX Demonstration Course
                </button>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                December 3, 2019
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              >
                <span
                  className="badge badge-light"
                >
                  Declined
                </span>
              </td>
              <td
                className="pgn__data-table-cell-wrap"
                role="cell"
              />
            </tr>
          </tbody>
        </table>
      </div>
      <div
        className="pgn__data-table-footer"
      >
        <div>
          Showing 
          3
           of 
          3
          .
        </div>
        <div
          className="btn-group btn-group-md"
          role="group"
        >
          <button
            aria-label="Previous page"
            className="btn-icon btn-icon-primary btn-icon-md"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <span
              className="btn-icon__icon-container"
            >
              <span
                className="pgn__icon btn-icon__icon"
                icon={Object {}}
              >
                <svg
                  aria-hidden={true}
                  fill="none"
                  focusable={false}
                  height={24}
                  role="img"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M17.51 3.87L15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </span>
          </button>
          <button
            aria-label="Next page"
            className="btn-icon btn-icon-primary btn-icon-md"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            <span
              className="btn-icon__icon-container"
            >
              <span
                className="pgn__icon btn-icon__icon"
                icon={Object {}}
              >
                <svg
                  aria-hidden={true}
                  fill="none"
                  focusable={false}
                  height={24}
                  role="img"
                  viewBox="0 0 24 24"
                  width={24}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M6.49 20.13l1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13z"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
